WebSharper.Owin/Owin.Sitelets.fs

```diff
@@ -81,35 +126,50 @@ module private Internal =
             | "trace" -> Http.Method.Trace
             | s -> Http.Method.Custom s
 
-        let Headers (headers: IHeaderDictionary) : seq<Http.Header> =
+        let Headers (headers: HeaderDictionary) : seq<Http.Header> =
             seq {
                 for KeyValue(k, vs) in headers do
                     for v in vs do
                         yield Http.Header.Custom k v
             }
 
-        let Query (query: IReadableStringCollection) : Http.ParameterCollection =
-            Http.ParameterCollection(
-                seq {
-                    for KeyValue(k, vs) in query do
-                        for v in vs do
-                            yield (k, v)
-                }
-            )
-
-        let Cookies (cookies: RequestCookieCollection) : HttpCookieCollection =
+        let Query (query: Arachne.Uri.Query) : Http.ParameterCollection =
+            match fst Arachne.Uri.Query.Pairs_ query with
+            | Some qs ->
+                Http.ParameterCollection(
+                    seq {
+                        for k, v in qs do
+                            yield k, match v with Some str -> str | None -> null
+                    }
+                )
+            | None -> Http.ParameterCollection(Seq.empty)
+
+        let tryFindCookieHeader (headers : HeaderDictionary) =
+            if headers.ContainsKey("Cookie") then
+                headers.["Cookie"]
+                |> Array.map Arachne.Http.State.Cookie.Parse

```

Comment:
Same here: https://github.com/freya-fs/arachne/blob/master/src/Arachne.Http.State/State.fs#L371 - bit confusing! Not too sure what might be happening here.